drivers/audio/wm8904.c

```diff
@@ -0,0 +1,759 @@
+	case 8000:
+		wm_sample_rate = kWM8904_SampleRate8kHz;
+		break;
+	case 12000:
+		wm_sample_rate = kWM8904_SampleRate12kHz;
+		break;
+	case 16000:
+		wm_sample_rate = kWM8904_SampleRate16kHz;
+		break;
+	case 24000:
+		wm_sample_rate = kWM8904_SampleRate24kHz;
+		break;
+	case 32000:
+		wm_sample_rate = kWM8904_SampleRate32kHz;
+		break;
+	case 48000:
+		wm_sample_rate = kWM8904_SampleRate48kHz;
+		break;
+	case 11025:
+		wm_sample_rate = kWM8904_SampleRate11025Hz;
+		break;
+	case 22050:
+		wm_sample_rate = kWM8904_SampleRate22050Hz;
+		break;
+	case 44100:
+		wm_sample_rate = kWM8904_SampleRate44100Hz;
+		break;
+	default:
+		LOG_WRN("Invalid codec sample rate: %d", cfg->i2s.frame_clk_freq);
+		return -EINVAL;
+	}
+
+	wm8904_read_reg(dev, WM8904_CLK_RATES_0, &mclkDiv);
+	fs = (mclk >> (mclkDiv & 0x1U)) / cfg->i2s.frame_clk_freq;
+
+	switch (fs) {
+	case 64:
+		wmfs_ratio = kWM8904_FsRatio64X;
+		break;
+	case 128:
+		wmfs_ratio = kWM8904_FsRatio128X;
+		break;
+	case 192:
+		wmfs_ratio = kWM8904_FsRatio192X;
+		break;
+	case 256:
+		wmfs_ratio = kWM8904_FsRatio256X;
+		break;
+	case 384:
+		wmfs_ratio = kWM8904_FsRatio384X;
+		break;
+	case 512:
+		wmfs_ratio = kWM8904_FsRatio512X;
+		break;
+	case 768:
+		wmfs_ratio = kWM8904_FsRatio768X;
+		break;
+	case 1024:
+		wmfs_ratio = kWM8904_FsRatio1024X;
+		break;
+	case 1408:
+		wmfs_ratio = kWM8904_FsRatio1408X;
+		break;
+	case 1536:
+		wmfs_ratio = kWM8904_FsRatio1536X;
+		break;
+	default:
+		LOG_WRN("Invalid Fs ratio: %d", fs);
+		return -EINVAL;
+	}
+
+	/* Disable SYSCLK */
+	wm8904_write_reg(dev, WM8904_CLK_RATES_2, 0x00);
+
+	/* Set Clock ratio and sample rate */
+	wm8904_write_reg(dev, WM8904_CLK_RATES_1,
+			 ((wmfs_ratio) << 10U) | (uint16_t)(wm_sample_rate));
+
+	switch (cfg->i2s.word_size) {
+	case 16:
+		word_size = 0;
+		break;
+	case 20:
+		word_size = 1;
+		break;
+	case 24:
+		word_size = 2;
+		break;
+	case 32:
+		word_size = 3;
+		break;
+	default:
+		LOG_ERR("Word size %d bits not supported; falling back to 16 bits",
+			cfg->i2s.word_size);
+		word_size = 0;
+		break;
+	}
+	/* Set bit resolution */
+	wm8904_update_reg(dev, WM8904_AUDIO_IF_1, (0x000CU), ((uint16_t)(word_size) << 2U));
+	{
+		uint16_t reg_val = 0;
+
+		wm8904_read_reg(dev, WM8904_AUDIO_IF_1, &reg_val);
+		LOG_INF("WM8904_AUDIO_IF_1 = %#x", reg_val);
+	}
+
+	/* Enable SYSCLK */
+	wm8904_write_reg(dev, WM8904_CLK_RATES_2, 0x1007);
+	return 0;
+}
+
+static int wm8904_out_volume_config(const struct device *dev, audio_channel_t channel, int volume)
+{
+	/**
+	 * WM8904_ANALOG_OUT1_LEFT, WM8904_ANALOG_OUT1_RIGHT (headphone outs),
+	 * WM8904_ANALOG_OUT2_LEFT and WM8904_ANALOG_OUT2_RIGHT (line outs):
+	 * [8]   - MUTE: leaving intact
+	 * [7]   - VU: 0 (values are "cached")
+	 * [6]   - ZC: 1 (zero-crossing enabled)
+	 * [5:0] - VOL: 6-bit volume value
+	 */
+
+	const uint16_t regval = (volume | 0b001000000) & 0b011111111;
+	const uint16_t mask = 0b011111111;
+
+	switch (channel) {
```

Comment:
This switch I would write it like:
```
uint8_t reg;
switch (channel) {
case AUDIO_CHANNEL_FRONT_LEFT:
	reg = WM8904_ANALOG_OUT2_LEFT;
	break;

case AUDIO_CHANNEL_FRONT_RIGHT:
	reg = WM8904_ANALOG_OUT2_RIGHT;
	break;

case AUDIO_CHANNEL_HEADPHONE_LEFT:
	reg = WM8904_ANALOG_OUT1_LEFT;
	break;

case AUDIO_CHANNEL_HEADPHONE_RIGHT:
	reg = WM8904_ANALOG_OUT1_RIGHT;
	break;

case AUDIO_CHANNEL_ALL:
	/* Set volume values with VU = 0 */
	reg = WM8904_ANALOG_OUT1_LEFT | WM8904_ANALOG_OUT1_RIGHT | WM8904_ANALOG_OUT2_LEFT | WM8904_ANALOG_OUT2_RIGHT;
	break;
default:
	return -EINVAL;
}

wm8904_update_reg(dev, reg, mask, regval);
return 0;
```

I see you do like this, here: https://github.com/zephyrproject-rtos/zephyr/pull/75943/commits/6f41b45b2320d72424dd58fe188eeb257da57228#diff-abe7424de1cefb3ba684964873bb5830e03f03ec4dd059ea2c67c07a8102548aR376

Reply:
That can't be done, as the `reg` variable holds an ordinal register index, it isn't a flag-style variable with disjunctive bits.

Reply:
ok, got it.